ls1intum/Artemis
# `General`, `Exam mode`: Modernize appearance of charts in exam statistics and exercise detail

### Checklist
#### General


#### Client

### Motivation and Context


This PR changes the implementations of the charts on the exam scores page and the doughnut charts on every exercise detail page from `ng2-charts` to `ngx-charts`.

### Description

`Ngx-charts` implementations now replace the `ng2-charts` implementations in the html templates. Furthermore, the underlying logic has been refactored in order to fit the new frameworks needs.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Exam in which students participated and that is already graded
- 1 arbitrary exercise **with release date in the past**

**Important**: Before deploying this PR, take a look at the old implementation on the test server. Make yourself familiar with the scores page. Taking screenshots might be helpful in order to compare the two implementations.

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to the exams overview of the corresponding course
4. For the exam, click on the scores button
5. Inspect the page, make sure that the information displayed by the charts has not changed. Also check the tooltips
6. Make sure that by clicking on horizontal bar of a specific exercise, you are forwarded to the scores page of the exercise like it was the case in the old implementation
7. Add/Modify the grading key and make sure that the changes are reflected by the vertical bar chart on the exam scores page
8. Go back to Course administration and enter the exercises page of the course
9. Choose an exercise with release date in the past and enter its detail page (by clicking on the name)
10. Inspect the three doughnut charts. Make sure that by clicking on the titles and parts of the charts the routing still works as with the old implementation.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





- exam-scores-average-scores-graph.component.ts: 62.16% | 82.6%
- exam-scores.component.ts: 76.57% | 94.78%
-  doughnut-chart.component.ts: 70% | 89.18% 

### Screenshots


Old implementation of vertical bar chart:
![image](https://user-images.githubusercontent.com/80622272/143684419-9401b5d2-d7ee-4537-be33-b0e1a8108db4.png)

*New* implementation of vertical bar chart:
![image](https://user-images.githubusercontent.com/80622272/143684574-83a6cd64-7079-474c-b7a7-c1b46dea7357.png)

Old implementation of horizontal bar charts:
![image](https://user-images.githubusercontent.com/80622272/143684434-5bdec02e-86b9-4fa1-8df9-bebe7fb50708.png)

*New* implementation of horizontal bar charts:
![image](https://user-images.githubusercontent.com/80622272/143685698-21acfefc-f53e-444e-bec1-8a1598abfff0.png)

Old implementation of doughnut charts on exercise detail page:
![image](https://user-images.githubusercontent.com/80622272/143860241-9f522b8f-780f-49d5-ac0e-8141545bd118.png)

*New* implementation of doughnut charts on exercise detail page:
![image](https://user-images.githubusercontent.com/80622272/143860590-de4f4658-a73e-44a1-90bc-6da496572690.png)
